File: js/3.js
let fs = require("fs")

const sum = (a, b) => a + b

const read = (str) => fs.readFileSync(str).toString().split("\n") // get data
const convToNum = (char) =>
  char.codePointAt(0) > 96 ? char.codePointAt(0) - 96 : char.codePointAt(0) - 38

const splitTwo = (str) => {
  const half = Math.floor(str.length / 2)
  return [str.slice(0, half), str.slice(half, str.length)]
}
const splitThree = (res, line, i) => {
  i % 3 !== 0 && res.length > 0 ? res[res.length - 1].push(line) : res.push([line])
  return res
}

const match = (args) => {
  let match
  Array.from(args[0]).forEach((char) => {
    if (args.slice(1).every((str) => [...str].includes(char))) {
      match = char
      return
    }
  })
  return match
}

const data = read("../src/3.txt")
const partOne = data.map(splitTwo).map(match).map(convToNum).reduce(sum, 0)
const partTwo = data.reduce(splitThree, []).map(match).map(convToNum).reduce(sum, 0)
console.log("Part 1 Sum:", partOne, "\n", "Part 2 Sum:", partTwo)

File: py/3.py
from functools import reduce


def read():
    f = open("../src/3.txt")
    return f.read().splitlines()


def convToNum(c):
    return ord(c[0]) - 96 if ord(c[0]) > 96 else ord(c[0]) - 38


def splitTwo(str):
    half = len(str)//2
    return [str[:half], str[half:len(str)]]


def splitThree(lines):
    res = []
    for i, line in enumerate(lines):
        if (i % 3 == 0 or len(res) == 0):
            res.append([line])
        else:
            res[len(res) - 1].append(line)

    return res


def match(*args):
    match = ""
    rest = args[0][1:len(args[0])]
    for char in args[0][0]:
        bool_list = []
        for ch in rest:
            if (char in ch):
                bool_list.append(True)
            else:
                bool_list.append(False)

        if (all(bool_list)):
            match = char
            break
    return match


data = read()


part1 = reduce(lambda a, b: a+b, map(convToNum,
                                     map(match, map(splitTwo, data))), 0)

part2 = reduce(lambda a, b: a+b, map(convToNum,
                                     map(match, splitThree(data))), 0)
print("Part 1 Score:", part1)
print("Part 2 Score:", part2)
